[VecOps] Add to ROOT the first implementation of VecOps

dpiparo opened the pull request:
https://github.com/dpiparo/VecOps/releases/tag/v0.1

bluehood commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,381 @@
+}
+
+template <typename T, typename F>
+TVec<typename TCallTraits<F>::ret_type> Operate(const TVec<T> &v0, F f)
```
wow this is `TDF::Map` implemented in #1612 !

dpiparo replied:
I apologise for having forgotten this.

dpiparo replied:
Operate is a terrible name to use instead of the widely understood *Map*

bluehood commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,381 @@
+}
+
+template <typename T, typename V>
+TVec<char> operator>(const TVec<T> &v0, const TVec<V> &v1)
```
I know `vector<bool>` is evil, but `vector<char>` looks very weird in a public signature.
maybe we can alias is to `TVec::Bool_t`, or maaaaaaaaybe just maybe we can actually use `vector<bool>`: what would break?

dpiparo replied:
This signature is a consequence of a more general design choice which consists in not using TVec<bool> but rather something which would not rely on the specialised implementation of the collection.
Let me tackle the questions backwards.
- The forthcoming buffer protocol implementation and all the migration of "contiguous blocks of memory" from ROOT to NumPy would fail, right?
- Wouldn't vectorisation performance be impacted by the usage of a vector<bool>?
- The usage of a type which seems to suggest it's a boolean but it's not might be more elegant but wouldn't it be even weirder than the signature with char?
Maybe a solution could really consist in encapsulating the present TVec implementation as TVecImpl in a new TVec template which could well be defined as TVec<bool> but would internally be implemented as a TVecImpl<char>.

bluehood replied:
afaik buffer protocol and vectorisation could work with `vector<bool>` as it still stores data in contiguous memory. it might make it awkward though, if it requires special handling.

anyway I was just wondering, but no matter, let's stay away from `vector<bool>` :)
what about `vector<int>` or `vector<short>` though? they might look less surprising when people read the signature.

bluehood replied:
I agree (see above), except it can't be auto-deduced because we don't want `vector<bool>` to pop up in bad places

amadio replied:
Indeed, I just checked and the comparison type is a `bool` in C++, not an `int`, but in hardware, using an integer type (either `int` or `long`, depending on the compared type) is probably best.

bluehood commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,381 @@
+}
+
+template <typename T, typename V>
+TVec<T> Filter(const TVec<T> &v, const TVec<V> &conds)
```
This is also in #1612 :)

dpiparo replied:
I am wondering if this is bad or good. Probably the two can coexist? in #1612 we have the signature which accommodates a callable which is more general.
I am not sure yet what is the right way to proceed here...

bluehood replied:
I'm for moving `Map` and `Filter` from #1612 over here -- or whatever you prefer! :)

bluehood replied:
`std::copy` and `std::copy_if` are c++11, the problem is that they have horrible signatures. imho `Filter` should wrap `copy_if` just as `Operate` here maps `std::transform` (see #1612 ), but I think @dpiparo is going to merge the two PRs or something like that

dpiparo replied:
The issue here is that copy_if does not allow to have, without hacks, iteration on 2 collections but rather 1 collection and a predicate.
Here the conditions according to which individual elements of the TVec have to be retained are packed in a TVec too.

amadio replied:
But packing in a `TVec` means that you have to loop once to compute the `conds`, and then a second time to copy things. If you use a lambda as predicate, then you can probably get away with a single loop, no?

bluehood replied:
yes, as things stand, `v[v>3]` loops twice, `Filter(v, [](int v) { return v > 3; })` loops once. both options will be available if I understand correctly, with obvious tradeoffs

bluehood commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,381 @@
+}
+
+template <typename T>
+TVec<T> Filter(const TVec<T> &v, bool cond)
```
In what case would we need this?

dpiparo replied:
We support for all operations on TVecs the 2 cases: scalar and vector. This is done to preserve symmetry. A case which comes to my mind is filtering a collection, say jets, based on a scalar quantity, say MET or median energy density.

dpiparo replied:
To comment again, I am not sure how one could implement this behaviour in C++ once we buy into the syntax **v[v < 3]** since that would imply the "wrong" overload to be picked up :)

amadio replied:
This could be accomplished if `v < 3` returned some special type, like `TMaskVec<T>`, which would mean you always get the right overload. This is what is done in [`std::valarray`](http://en.cppreference.com/w/cpp/numeric/valarray/operator_at).

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,381 @@
+}
+
+template <typename T, typename V, typename F>
+TVec<typename TCallTraits<F>::ret_type> Operate(const TVec<T> &v0, const TVec<V> &v1, const char* opName, F f)
```
Fine!

amadio commented on tree/treeplayer/inc/ROOT/TVecAllocator.hxx:
```diff
@@ -0,0 +1,88 @@
+
+   pointer allocate(std::size_t n)
+   {
+      if (n > std::size_t(-1) / sizeof(T))
```
This will lead to warnings about mixing up negative values with unsigned types. Also, why throw here instead of leaving that to the real allocator? And why ignore `n` when reusing `fInitialAddress`? Will this not let people think they allocated more space when in fact they reused a pointer that has lower capacity?

bluehood replied:
I don't think this prints any warnings, the comparison is between `std::size_t` and `std::size_t(-1)` is allowed afaik. That check comes from [cppreference](http://en.cppreference.com/w/cpp/concept/Allocator)'s demo allocator implementation by the way.

amadio commented on tree/treeplayer/test/dataframe/vecops_tvec.cxx:
```diff
@@ -0,0 +1,175 @@
+   asinh(ref);
+   acosh(ref);
+   tanh(ref);
+   atanh(ref);
```
Shouldn't this check the results too? Is `ref2` unused?

amadio commented on tree/treeplayer/test/dataframe/vecops_tvecallocator.cxx:
```diff
@@ -0,0 +1,77 @@
+
+   auto res = 0;
+   try {
+      v.resize(-1);
```
Please use a standard thing like `std::numeric_limits<size_t>::max()` or `INT64_MAX` or, if you must, use `~0UL` instead of `-1` (note that `-1` may evaluate to `~0U == 4294967295`, not `~0UL == 18446744073709551615`, which is likely what you want here).

dpiparo replied:
I agree, thanks.

bluehood replied:
I agree with using something more explicit, but the code as is is not ambiguous, you will get a vector with size `std::size_t(-1)` which is well defined

bluehood commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,381 @@
+
+namespace Detail {
+
+namespace VecOps {
```
we had to take this decision for TDF...I don't remember the rationale for `ROOT::Internal::TDF` rather than `ROOT::TDF::Internal` but it was preferred -- I think @pcanal argued for the way we did it at the time

dpiparo replied:
@pcanal has a point. I would, if possible, avoid proliferation of namespaces and keep a clear hierarchy. For example, inside ROOT::Internal, there are several families of "internal" routines and helper classes, each family labelled by a namespace.

amadio replied:
The number of namespaces seems to be the same to me in both strategies. One just makes code more verbose because you always have to start names from the top namespace...

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,381 @@
+#define MATH_FUNC(FUNCNAME)\
+template <typename T>\
+auto FUNCNAME(const TVec<T> &v) -> TVec<decltype(std::FUNCNAME(v[0]))>\
+{return ROOT::Internal::VecOps::Operate(v, [](const T& t) {return std::FUNCNAME(t);});}
```
of course. For those, especially TMath's, we might want to have something dedicated, also depending on where the VecOps will land in the end. The current location, TreePlayer, is a bit awkward.

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,381 @@
+   return ROOT::Internal::VecOps::Operate(v, [&c](const T& t) {return t/c;});
+}
+
+template <typename T, typename V, typename D = typename std::enable_if<!ROOT::Detail::VecOps::TIsTVec<V>::value, int>::type>
```
I acknowledge all comments about the usage of char as boolean type, both coming from the programming model point of view and the optimisation one. This will be changed.

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,381 @@
+{
+   CheckSizes(v0.size(), v1.size(), opName);
+   TVec<typename TCallTraits<F>::ret_type> w;
+   w.resize(v0.size());
```
Absolutely.

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,381 @@
+template <typename T, typename V, typename F>
+TVec<typename TCallTraits<F>::ret_type> Operate(const TVec<T> &v0, const TVec<V> &v1, const char* opName, F f)
+{
+   CheckSizes(v0.size(), v1.size(), opName);
```
But would not this imply massive copy&paste and more error prone code?

amadio replied:
It's up to you. Since this is about performance, maybe avoiding the extra work to pass parameters would be worth it.

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,381 @@
+      err << "Cannot perform operation " << opName
+          << ". The array sizes differ: "
+          << s0 << " and " << s1 << std::endl;
+      throw std::runtime_error(err.str());
```
sure. Done.

dpiparo commented on tree/treeplayer/inc/ROOT/TVecAllocator.hxx:
```diff
@@ -0,0 +1,88 @@
+namespace VecOps {
+
+template <typename T>
+class TVecAllocator // : public std::allocator<T>
```
sure. Done.

dpiparo commented on tree/treeplayer/inc/ROOT/TVecAllocator.hxx:
```diff
@@ -0,0 +1,88 @@
+         throw std::bad_alloc();
+      if (EAllocType::kNoneYet == fAllocType) {
+         fAllocType = EAllocType::kFromExternalPointer;
+         return (pointer)fInitialAddress;
```
Sure, you do have a point.

dpiparo commented on tree/treeplayer/inc/ROOT/TVecAllocator.hxx:
```diff
@@ -0,0 +1,88 @@
+   {
+      if (n > std::size_t(-1) / sizeof(T))
+         throw std::bad_alloc();
+      if (EAllocType::kNoneYet == fAllocType) {
```
I may misunderstand your comment, but I think the logic is the following one:
1) If the allocator is initialised in any other way, its state is kRegular
2) If the allocator is initialised with an address and a size, its state is kNoneYet, which we may call kFromExternalPointerNoAllocateCalledYet

If 1) we just behave like a normal allocator.
If 2) we need to return the initial address the first time. At the second time we need to flip the state to kRegular and, from that call on, *behave normally*. It's not clear to me how you can *behave normally* without having a 3rd state which prevents to return steadily fInitialAddress.

But maybe, **I** am missing something :)

dpiparo replied:
what happens in this scenario if construct is invoked after allocate and a pointer was passed?

bluehood replied:
indeed bad things happen :)

dpiparo commented on tree/treeplayer/inc/ROOT/TVecAllocator.hxx:
```diff
@@ -0,0 +1,88 @@
+   void construct(pointer p, const_reference val)
+   {
+      if (EAllocType::kFromExternalPointer == fAllocType)
+         return;
```
I was admittedly neglecting doc to write it properly once functionality was achieved (the creation of VecOps was done exploring many strategies), but you are right. This is a subtle point.

dpiparo commented:
Dear @amadio, @bluehood ,
thanks for the complete, smart and constructive set of comments. I think I implemented them all. 
In addition, I reloaded the implementation of the **VecOps**. In a nutshell the changes are:
- Now TVec encapsulate a vector<T, TVecAllocator<T>>. This allows to mirror the interface of vector but also to allow nice goodies like the constructor from an address and a lenght or assignment ops.
- Map free function has been added to do what Python's map does
- Filter was not added but rather the operator[] was overloaded to allow things like **v[ v < 3 ]** like numpy does
- Assignment operators have been added
- Test coverage has been increased

I wonder if this, once TVec replaces TArrayBranch in the TDF implementation, is enough to make life of users like @JavierCVilla .
I added @stwunsch as reviewer given the chat we had face2face and his involvement in analysis and TMVA development.

bluehood commented on tree/treeplayer/test/dataframe/vecops_tvec.cxx:
```diff
@@ -0,0 +1,238 @@
+   EXPECT_EQ(wsize, v.size());
+   for (unsigned int i = 0; i < wsize; i++) {
+      EXPECT_DOUBLE_EQ(v[i], w[i]) << " error while checking math function " << name;
+   }
```
seems like we are missing `All` and `Any`: this should be
```c++
EXPECT_TRUE(All(v[i] == w[i]));
```
or to be super correct `AllClose(v[i] == w[i])`.

dpiparo replied:
Sorry I do not get this: what do you mean?

bluehood replied:
the fact that you had to write 4 lines to check that all elements of one TVec were equal to all elements of another shows that we are missing some useful helpers: `All`, `Any` and `AllClose` (stealing the numpy denominations)

dpiparo replied:
Oh, now I see. Fair enough.

dpiparo replied:
std::equal is probably to be used, or a wrapper around it.

bluehood commented on tree/treeplayer/inc/ROOT/TVecAllocator.hxx:
```diff
@@ -0,0 +1,89 @@
+namespace VecOps {
+
+template <typename T>
+class TVecAllocator {
```
imho this allocator is of general interest and applicability, and is not bound to be used in conjunction with `TVec`, so I would give it a name that reflects what it does, like `TAdoptAllocator` or `TViewAllocator`

dpiparo replied:
Good point.
At this point we could move it out from Detail::VecOps and put it somewhere else... Not so sure this place is core/cont.

dpiparo replied:
Almost addressed. The new name TAdoptAllocator has been selected, the code documented, basically we are ready to move it to ROOT::Experimental if we want.

dpiparo commented on tree/treeplayer/test/dataframe/vecops_tvec.cxx:
```diff
@@ -0,0 +1,238 @@
+   ROOT::Experimental::VecOps::TVec<int> w(ref.data(), ref.size());
+
+   CompAndPrintTVec(v, 2.);
+   CompAndPrintTVec(v, ref + 2);
```
This was a macro translated to a test. We need to capture, you are right.

dpiparo replied:
Addressed.

dpiparo commented on tree/treeplayer/test/dataframe/vecops_tvec.cxx:
```diff
@@ -0,0 +1,238 @@
+template <typename T, typename V>
+void CompAndPrintTVec(ROOT::Experimental::VecOps::TVec<T> v, V w)
+{
+   using namespace ROOT::Experimental::VecOps;
```
I think at some point I needed the line, easy to check :) Thanks!

dpiparo replied:
Addressed (it is still necessary, at least on osx :) )

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,535 @@
+{
+   return ROOT::Internal::VecOps::Operate(v, f);
+}
+
```
I was aggressively removing entities which could be expressed otherwise. Given the cost, we can keep the aforementioned helper instead of a syntax like v[Map(v, unaryOp)]...

dpiparo replied:
Addressed.

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,535 @@
+}
+
+template <typename T, typename V, typename F>
+TVec<typename TCallTraits<F>::ret_type> Operate(const TVec<T> &v0, const TVec<V> &v1, std::string_view opName, F f)
```
I agree with the suggestion.

dpiparo replied:
Addressed.

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,535 @@
+
+using namespace ROOT::Experimental::VecOps;
+
+void CheckSizes(size_t s0, size_t s1, std::string_view opName)
```
For me it's fine. Why wouldn't the current implementation be inlined? What would you exactly do to improve performance?
Another option is to have an assert and just check this in dbg mode...

bluehood replied:
> What would you exactly do to improve performance?
my suggestion is not to optimize this, but that we remind ourselves to check how much impact on runtime these `CheckSizes` calls have. Using `assert` as you mentioned could be an alternative if we see this is a problem.

> Why wouldn't the current implementation be inlined?
It's very hard to guess what compilers inline, but for example building a stringstream with runtime values might be a deterrent

My point is just that given that `CheckSizes` appears in many functions that are supposed to be "fast", we should check

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,535 @@
+class TVec;
+
+template <typename T>
+using TCallTraits = typename ROOT::TypeTraits::CallableTraits<T>;
```
Definitively...

dpiparo replied:
To be seen if this will be still necessary. See your comment about decltype below!

dpiparo replied:
addressed.

dpiparo commented:
@bluehood , thanks a lot for investing time in the comments. They are not pedantic and more than 1 pair of eyes is needed for something like VecOps to be "right" both in terms of implementation and, above all, interface.

bluehood commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,535 @@
+}
+
+template <typename T, typename V>
+TVec<Boolean_t> operator==(const TVec<T> &v0, const TVec<V> &v1)
```
p.s. I understand that it's desirable for `operator==` to behave like this, I was just thinking out loud here...

dpiparo replied:
you have a point. Here I do not know really how to proceed. I think that in case of TVec  and therewith analysis `operator==` should really behave as it is doing now. TVec is born to help analysis. If a "proxying vector" is needed, one can directly use vector<T, TAdoptAllocator<T>>, right?

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,535 @@
+
+using namespace ROOT::Experimental::VecOps;
+
+void CheckSizes(size_t s0, size_t s1, std::string_view opName)
```
addressed.

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,535 @@
+}
+
+template <typename T, typename V, typename F>
+TVec<typename TCallTraits<F>::ret_type> Operate(const TVec<T> &v0, const TVec<V> &v1, std::string_view opName, F f)
```
Addressed.

bluehood commented:
hi @dpiparo, that's amazing! of the remaining comments I think the really important ones are those regarding a missing reference in `const&` and the missing defaulted copy- and move-ctors for `TVec`, the rest is much less than critical.

bluehood commented:
uhm...it might be a problem if osx users cannot use operator<< without
explicitly bringing it in scope

On 14 Feb 2018 9:28 pm, "Danilo Piparo" <notifications@github.com> wrote:

> *@dpiparo* commented on this pull request.
> ------------------------------
>
> In tree/treeplayer/test/dataframe/vecops_tvec.cxx
> <https://github.com/root-project/root/pull/1613#discussion_r168300519>:
>
> > +   using namespace ROOT::Experimental::VecOps;
> +   TVec<int> v{0, 1, 2, 3, 4, 5};
> +   std::vector<int> vEvenRef{0, 2, 4};
> +   std::vector<int> vOddRef{1, 3, 5};
> +   auto vEven = v[v % 2 == 0];
> +   auto vOdd = v[v % 2 == 1];
> +   for (int i = 0; i < 3; ++i) {
> +      EXPECT_EQ(vEven[i], vEvenRef[i]);
> +      EXPECT_EQ(vOdd[i], vOddRef[i]);
> +   }
> +}
> +
> +template <typename T, typename V>
> +void CompAndPrintTVec(ROOT::Experimental::VecOps::TVec<T> v, V w)
> +{
> +   using namespace ROOT::Experimental::VecOps;
>
> Addressed (it is still necessary, at least on osx :) )
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/root-project/root/pull/1613#discussion_r168300519>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKfU-gfWThw-ZO5RbdJmIoax9sKgtj-2ks5tU0H4gaJpZM4SAHX->
> .
>

bluehood commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,536 @@
+   TVec(std::initializer_list<T> init) : fData(init) {}
+   TVec(pointer p, size_type n) : fData(n, T(), ROOT::Detail::VecOps::TAdoptAllocator<T>(p, n)) {}
+   // assignment
+   TVec<T> &operator=(std::initializer_list<T> ilist) { return fData = ilist; }
```
`operator=(const TVec&)` and `operator=(TVec &&)` are missing

dpiparo replied:
I left all of the package intentionally aside to have them defaulted: the only data member is an stl vector. Do you see a particular need for implementing them?

bluehood replied:
It would be better to explicitly declare them as `default` to make sure that they will not be silently deleted by future changes to the library. Move-assignment is particularly tricky, because if deleted C++ defaults to use the copy-assignment, with silent loss in performance (hard to catch even in unit tests).

dpiparo replied:
sorry. See above :(

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,535 @@
+   using reverse_iterator = typename Impl_t::reverse_iterator;
+
+   // ctors
+   TVec() {}
```
Fair enough.

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,535 @@
+template <typename T>
+struct TIsChar {
+   static constexpr bool value =
+      TIsOneOf<typename std::decay<T>::type, char, signed char, unsigned char, wchar_t, char16_t, char32_t>::value;
```
yes, this one might be simplified.

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,535 @@
+   void pop_back() { fData.pop_back(); }
+   void resize(size_type count) { fData.resize(count); }
+   void resize(size_type count, const value_type &value) { fData.resize(count, value); }
+   void swap(TVec<T> &other) { std::swap(fData, other.fData); }
```
oh yes! Thanks for the reminder!

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,535 @@
+inline std::string printValue(ROOT::Experimental::VecOps::TVec<T> *tvec)
+{
+   std::stringstream ss;
+   ss << *tvec;
```
I think so...

Axel-Naumann replied:
Yes it's guaranteed (this is just because it's a lot easier to implement overloading / fallback etc)

bluehood replied:
`gsl::not_null` ftw :)

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,535 @@
+template <typename T>
+T Sum(const TVec<T> v)
+{
+   return std::accumulate(v.begin(), v.end(), 0);
```
nice :)

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,535 @@
+
+template <typename T, typename V,
+          typename D = typename std::enable_if<!ROOT::Detail::VecOps::TIsTVec<V>::value, int>::type>
+TVec<Boolean_t> operator>(const TVec<T> &v, const V &c)
```
this works well in ProcessExecutor, it's to be tried indeed, the functionality is the same.

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,535 @@
+      const auto thisSize = size();
+      ROOT::Internal::VecOps::CheckSizes(thisSize, conds.size(), "operator[]");
+      TVec<T> w;
+      w.reserve(thisSize);
```
Here we may decide for something else, but would the initialisation of the memory be less expensive than a "generous" reserve?

bluehood replied:
I don't know 😅

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,535 @@
+std::ostream &operator<<(std::ostream &os, const TVec<T> &v)
+{
+   // In order to print properly, convert to 64 bit int if this is a char
+   using Print_t = typename std::conditional<ROOT::Internal::VecOps::TIsChar<T>::value, long long int, T>::type;
```
I would still do something. If a TVec<char> is used, then printing is broken. We can certainly refactor TIsChar...

bluehood commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -484,7 +468,13 @@ template <class T>
+                                std::is_same<wchar_t, T>::value ||
+                                std::is_same<char16_t, T>::value ||
+                                std::is_same<char32_t, T>::value;
+   using Print_t = typename std::conditional<mustConvert, long long int ,T>::type;
```
why do we need to convert `char` to `long long int` when printing the contents of `TVec<char>`?

dpiparo replied:
Because we assume these chars are numbers (VecOps are mainly intended for analysis and not text manipulation) and this is a comfortable way to print the value of these numbers.

bluehood replied:
ok, in other words we assume that physicists/experiments predominantly use arrays of chars as memory-efficient arrays of short integers, and print them as such.

dpiparo commented:
@phsft-bot build please.

dpiparo commented:
@phsft-bot build please

Axel-Naumann commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,574 @@
+      return ROOT::Internal::VecOps::OperateInPlace(*this, v0, "%", [](const T &t, const V &v) { return t % v; });
+   }
+};
+
```
Do we want a ternary `cond ? a : b`?

dpiparo replied:
Do you mean something that loops over groups of three elements from 3 TVecs and returns a TVec with the result of a ternary applied to the 3? If yes, would this be another free function (i.e. can we really overload "?")?

Axel-Naumann commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,574 @@
+*/
+///@{
+template <typename T, typename V>
+auto operator+(const TVec<T> &v, const V &c) -> TVec<decltype(v[0] + c)>
```
Could you make them friends of TVec, available only for ADL-lookup?

dpiparo replied:
Yes, thanks for the suggestion. This would make thinks much easier for the user.

Axel-Naumann commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,574 @@
+   return ROOT::Internal::VecOps::Operate(v, [&c](const T &t) { return t % c; });
+}
+
+using Boolean_t = int;
```
?! Probably worth at least a comment...

dpiparo replied:
Sure, thanks for the observation. The problem behind this alias is to avoid the specialisation vector<bool>

Axel-Naumann commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,574 @@
+using Boolean_t = int;
+
+template <typename T, typename V>
+auto operator>(const TVec<T> &v, const V &c) -> decltype(v[0] > c, TVec<Boolean_t>())
```
What about `op>(V, TVec<T>)`? (Possibly also for all the others.)

dpiparo replied:
Good point. It's odd to implement non-commutative arithmetic ;-)

Axel-Naumann commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,574 @@
+MATH_FUNC(tanh)
+MATH_FUNC(atanh)
+MATH_FUNC(abs)
+
```
`#undef MATH_FUNC`?

dpiparo replied:
:( of course.

Axel-Naumann commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,574 @@
+
+/// Create a new collection with the elements passing the filter expressed by the predicate
+template <typename T, typename F>
+TVec<T> Filter(const TVec<T> &v, F &&f)
```
Should we also have an operation that uses a mask?

dpiparo replied:
Isn't this &&? For example: mu_pt && (abs(mu_eta) < 2)

Axel-Naumann replied:
Isn't that *resulting* in a mask? What if I have such a mask and I'd like a vector that is possibly smaller, selecting only those elements from the original vector that pass the mask? I.e. in `Filter()`s terms:
```
auto mask = Map(v, f);
assert(mask.size() == f.size());
auto reduced = ThisOpMightBeMissing_CouldBeCalledReduce(v, mask);
```
instead of
```
auto reduced = Filter(v, f);
```

dpiparo replied:
I see (with 48h delay, sorry).
This is what is available with the overloaded operator[], like in numpy:
```
auto reduced = v[mask];
```
or with a Physics hat on top:
```
auto goodPts = pts[pts > 15 && abs(etas) < 1.2];
```

Axel-Naumann commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,574 @@
+
+public:
+   // ctors
+   TVec() {}
```
It'd be *really* good to have at least a move ctor and maybe even a copy ctor.
Same with assignment.

dpiparo replied:
You are right. And @bluehood  also note this. The change I had got lost in a rebase: this will be of course addressed. Sorry about that!

bluehood commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,574 @@
+#ifndef ROOT_TVEC
+#define ROOT_TVEC
+
+#include "RStringView.h"
```
I left `RStringView.h` there for backwards compatibility...

Axel-Naumann replied:
OK let me rephrase then: shouldn't we `#include <ROOT/RStringView.hxx>`?

dpiparo replied:
Yes he did yesterday and I did not rebase, sorry!

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,574 @@
+   return w;
+}
+
+template <typename T, typename V, typename F>
```
Fair enough.

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,574 @@
+
+public:
+   // ctors
+   TVec() {}
```
:) Yes...

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,574 @@
+   explicit TVec(size_type count) : fData(count) {}
+   TVec(const std::vector<T> &other) { std::copy(other.begin(), other.end(), fData.begin()); }
+   TVec(std::initializer_list<T> init) : fData(init) {}
+   TVec(pointer p, size_type n) : fData(n, T(), ROOT::Detail::VecOps::TAdoptAllocator<T>(p, n)) {}
```
But this is not exactly transfer of ownership, right? Here we are adopting the memory pointed by "p".

Axel-Naumann replied:
I see, the memory pointed to by `p` will never be deleted by the allocator... Indeed, then ignore my comment, please.

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,574 @@
+      w.reserve(thisSize);
+      for (std::size_t i = 0; i < thisSize; i++) {
+         if (conds[i])
+            w.emplace_back(fData[i]);
```
I think spelling the predicate is not an issue. What I was trying to do is not to initialise the memory associated to w but rather reserve it not to pay the price of initialisation.

dpiparo replied:
Oh, but maybe this could be handled by a back_inserter...

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,574 @@
+   template <typename V>
+   TVec<T> &operator+=(const V &c)
+   {
+      return ROOT::Internal::VecOps::OperateInPlace(*this, [&c](const T &t) { return t + c; });
```
MMhh, good point. A the end of the day that's what this free function really is... I left it around after a refactoring without realising this!

dpiparo commented on tree/treeplayer/inc/ROOT/TVec.hxx:
```diff
@@ -0,0 +1,574 @@
+   template <typename V>
+   TVec<T> &operator+=(const TVec<V> &v0)
+   {
+      return ROOT::Internal::VecOps::OperateInPlace(*this, v0, "+", [](const T &t, const V &v) { return t + v; });
```
ops :(

dpiparo replied:
uhm.. What is the difference? I'd prefer to keep a const t and not use the += op in the lambda.

dpiparo commented on tree/treeplayer/test/dataframe/vecops_tvec.cxx:
```diff
@@ -0,0 +1,238 @@
+TEST(VecOps, MoveCtor)
+{
+   ROOT::Experimental::VecOps::TVec<int> v1{1, 2, 3};
+   ROOT::Experimental::VecOps::TVec<int> v2(std::move(v1));
```
Thanks for the reminder. Indeed the default moves, ctors, assignments will be implemented (for real this time!). See above.

dpiparo commented:
All comments but ADL-lookup implemented (conceptually not difficult, just a big number of slides to be shuffled around), starting a round of builds.
